lambdaclass/mirra_backend
# [GH-525] Track match data

## Motivation
Closes #525 

We have to main reasons to track things that happen in-game during a match
1. Statistics and analysis for us (e.g. Why are people picking so much Muflus? Do the spam only certain skills? etc)
2. Post-match bonuses (e.g. daily quests, currency gains, etc)

## Summary of changes

Main change here is the introduction of `game_tracker.ex` this module will act as a collector of events, in a way its architected similar to how metrics/telemetry works and how their reporters works. Then is a matter of `game_updater.ex` emitting the events and properly tracking them, for now we are only emitting a kill event that keep tracks of the characters  you have killed and the character that killed you (saving your resulting position as well)

For this current implementation several decisions have been made that we I would like to make clear or talk about
- `GameTracker` is a single gen_server handling the tracking of all matches in the server, this might not scale, but its easily solvable by either having a pool of trackers or by having a dedicated tracker per match. Let's measure first before changing this
- Error handling for reporting of match results we should have a retry mechanism to prevent players from loosing game results, eventually this could be some "persistent" storage locally or another DB directly connected. For now this will be addressed in a follow-up #601 

## How to test it?

To test you can run a game through the browser (`game_client`) by going to [localhost:3000](http://localhost:3000/) and in the console see the result after the match ends

```
GameBackend.Repo.all(GameBackend.Matches.ArenaMatchResult)
```

## Checklist
